### PR TITLE
Implement two smoke tests for non-default settings

### LINKED ISSUE #4954: Implement more test cases
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Current Behavior

Command:
```
# no issue when USE_FIND_ALGORITHM=false
docker run --rm \
    -e RUN_LOCAL=true \
    -e USE_FIND_ALGORITHM=true \
    -v "$(git rev-parse --show-toplevel)":/tmp/lint ghcr.io/super-linter/super-linter:slim-latest
```

Error
```
...
[DEBUG] Validate the local Git environment
[DEBUG] Check if /tmp/lint is a Git repository
.git
[DEBUG] /tmp/lint is a Git repository
[DEBUG] Git branches: 
* main
  dev
[DEBUG] Validate that the GITHUB_SHA reference () exists in this Git repository.
[FATAL] The GITHUB_SHA reference () doesn't exist in this Git repository
```

5a8805dc4fb3ff22315a0c7f8ce8d3cc7da81958 is the first bad commit

### Expected Behavior

up to v5.7.2
Super-linter shall run till the end and report the result for both `USE_FIND_ALGORITHM=true` and `USE_FIND_ALGORITHM=false`

### Super-Linter version

```markdown
slim-latest
```

### Relevant log output

```shell
Digest of git bisect log


grep -i -E -e '(BUILD_REVISION|naming to|fatal)' bisect.log 
        --build-arg BUILD_REVISION=2ac4d451 \
#113 naming to ghcr.io/super-linter/super-linter:latest 0.0s done
fatal: Not a valid object name 
2023-12-06 06:10:27 [FATAL]   The GITHUB_SHA reference () doesn't exist in this Git repository
        --build-arg BUILD_REVISION=4645b766 \
#113 naming to ghcr.io/super-linter/super-linter:latest 0.0s done
fatal: Not a valid object name 
2023-12-06 06:12:48 [FATAL]   The GITHUB_SHA reference () doesn't exist in this Git repository
        --build-arg BUILD_REVISION=879672e9 \
#113 naming to ghcr.io/super-linter/super-linter:latest 0.0s done
fatal: Not a valid object name 
2023-12-06 06:14:35 [FATAL]   The GITHUB_SHA reference () doesn't exist in this Git repository
        --build-arg BUILD_REVISION=5a8805dc \
#113 naming to ghcr.io/super-linter/super-linter:latest 0.0s done
fatal: Not a valid object name 
2023-12-06 06:16:46 [FATAL]   The GITHUB_SHA reference () doesn't exist in this Git repository
5a8805dc4fb3ff22315a0c7f8ce8d3cc7da81958 is the first bad commit
```

### Steps To Reproduce

```
docker run --rm \
    -e RUN_LOCAL=true \
    -e USE_FIND_ALGORITHM=true \
    -v "$(git rev-parse --show-toplevel)":/tmp/lint ghcr.io/super-linter/super-linter:slim-latest
```

### Anything else?

The error message above is from here

https://github.com/super-linter/super-linter/blob/5a8805dc4fb3ff22315a0c7f8ce8d3cc7da81958/lib/functions/validation.sh#L237-L244

I assume this is for checking `GITHUB_SHA` provided by github-action against the checked out commit

In local run I have not seen documentation or reason why this variable has to be set, I guess user are more likely commanding `RUN_LOCAL=true` to
> lint my directory here as is, regardless it has git or not, it is dirty or not, or which reference, branch or tag I am at

supported by #4952 that a local run is used to lint a non-git directory.

I have not enough understanding on the commit 5a8805dc4fb3ff22315a0c7f8ce8d3cc7da81958 to move the issue forward or digging deeper down.

**Suggestion**

The only suggestion I would make is, as `RUN_LOCAL=true` is a supported and an important feature to guard non-conforming code from being pushed to remote, CI/CD should make better coverage on various global flags, and also the case of non-git directory.

| **ENV VAR**                   | **Default Value** | **Test Coverage and Requirement**  | **Notes**                                                                                                                                                                        |
| ----------------------------- | ----------------- | ---------------------------------- | -------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- |
| **ERROR_ON_MISSING_EXEC_BIT** | `false`           | `true` covered, is sufficient      | If set to `false`, the `bash-exec` linter will report a warning if a shell script is not executable. If set to `true`, the `bash-exec` linter will report an error instead.      |
| **USE_FIND_ALGORITHM**        | `false`           | `true` not covered, both required  | By default, we use `git diff` to find all files in the workspace and what has been updated, this would enable the Linux `find` method instead to find all files to lint          |
| **EXPERIMENTAL_BATCH_WORKER** | `false`           | `true` not covered, both required  | Flag to enable experimental parallel and batched worker. As of current only `eslint` and `cfn-lint` are supported, if there is no support, original version is used as fallback  |
| **IGNORE_GENERATED_FILES**    | `false`           | `true` not covered, both required  | If set to `true`, super-linter will ignore all the files with `@generated` marker but without `@not-generated` marker.                                                           |
| **IGNORE_GITIGNORED_FILES**   | `false`           | `true` not covered, both required  | If set to `true`, super-linter will ignore all the files that are ignored by Git.                                                                                                |
| **VALIDATE_ALL_CODEBASE**     | `true`            | `false` not covered, both required | Will parse the entire repository and find all files to validate across all types. **NOTE:** When set to `false`, only **new** or **edited** files will be parsed for validation. |
| **[non-env] Input directory is a git directory** |  no default                 | `false` not covered, both required | User might lint a git tracked directory, or a plain directory without version control, for some reason                                                                           |

### PR DESCRIPTION
**Pull request by kftsehk**
Fixes #4954 

## Proposed Changes

**Scope**

Guarentees super-linter builds can complete linting for local run in various non-default scenarios

Implements the suggestion in #4954 to increase test coverage for `RUN_LOCAL=true` for non-default general settings into ci/cd pipeline
- Non-default Workspace: `DEFAULT_WORKSPACE`
- Git Related Env: `IGNORE_GENERATED_FILES`, `IGNORE_GITIGNORED_FILES`, `VALIDATE_ALL_CODEBASE`
- Find Algorithm: `USE_FIND_ALGORITHM`
- Non Git-tracked Directory: directory without a `.git` folder
- Experimental Features: `EXPERIMENTAL_BATCH_WORKER`

**Out of scope**

There is NO assertion for correctness of the implementation, for example
- it does not assert `find` command is used instead of `git diff` when `USE_FILE_ALGORITHM=true`
- it does not assert `.gitignored` files are not linted when `IGNORE_GITIGNORED_FILES=true`
- likewise for others

These shall be tested as part of test suite or in other tests if necessary.

### Author/Contributor

- [x] I included all the needed documentation for this change.
- [x] I provided the necessary tests.

### Reviewing Maintainer

- [x] Label as `breaking` if this is a large, fundamental change.
- [x] Label as either: `automation`, `bug`, `documentation`, `enhancement`, `infrastructure`.

### CODE REVIEW

**Comment by kftsehk** [comment]
It is an expected behavior to have failing CI/CD before issues such as #4952 #4954 are fixed, as the tests are designed to run code-paths previously not tested, and involved in the above issues.

**Comment by ferrarimarco** [comment]
Thanks for this PR!

While we iterate on this one, can you remove the step to run `USE_FIND_ALGORITHM=true`? I'm taking care of it in #4953

Thanks!

**Comment by kftsehk** [comment]
seem to conflict with migration of #4982

**Review by ferrarimarco — CHANGES_REQUESTED** [review]
@kftsehk In #4982, we're refactoring the build process to reduce duplication. Can you kindly refactor this PR my moving the logic to run these additional test cases to the Makefile? Hopefully it's not going to be too much work :)

Thanks a lot!

**Comment by kftsehk** [comment]
sure will wait for your PR go through and update it

**Comment by kftsehk** [comment]
pipeline failed by unrelated issue:
jscpd liniting on `.github/workflows/cd.yml` and report failure issue creation as duplicated code

File: Makefile [hunk 1]
```diff
@@ -127,7 +128,33 @@ test-find: ## Run super-linter on a subdirectory with USE_FIND_ALGORITHM=true
 		-e ENABLE_GITHUB_ACTIONS_GROUP_TITLE=true \
 		-e DEFAULT_BRANCH=main \
 		-e USE_FIND_ALGORITHM=true \
-		-v "$(CURDIR)/.github":/tmp/lint/.github \
```
Comment (ferrarimarco) [inline Makefile:130]: This is needed, otherwise super-linter will not find rules files in `.github/linters`.

File: Makefile [hunk 2]
```diff
@@ -127,7 +128,33 @@ test-find: ## Run super-linter on a subdirectory with USE_FIND_ALGORITHM=true
...
+.phony: test-non-default-workdir
+test-non-default-workdir: ## Run super-linter with DEFAULT_WORKSPACE set
+	docker run \
+		-e RUN_LOCAL=true \
+		-e ACTIONS_RUNNER_DEBUG=true \
+		-e ERROR_ON_MISSING_EXEC_BIT=true \
+		-e ENABLE_GITHUB_ACTIONS_GROUP_TITLE=true \
+		-e DEFAULT_BRANCH=main \
+		-e DEFAULT_WORKSPACE=/tmp/not-default-workspace \
+		-v "$(CURDIR)/.github:/tmp/not-default-workspace" \
```
Comment (ferrarimarco) [inline Makefile:143]: Refactoring this to `-v $(CURDIR)/.github:/tmp/not-default-workspace/.github` should fix the issues with JSCPD.

**Review by ferrarimarco — CHANGES_REQUESTED** [review]

File: Makefile (hunk 1 above)
Comment (ferrarimarco) [reply Makefile:130]: Can you please revert this change?

File: Makefile [hunk 3]
```diff
@@ -127,7 +128,33 @@ test-find: ## Run super-linter on a subdirectory with USE_FIND_ALGORITHM=true
...
+test-git-flags: ## Run super-linter with different git-related flags
+	docker run \
+		-e RUN_LOCAL=true \
+		-e ACTIONS_RUNNER_DEBUG=true \
+		-e ERROR_ON_MISSING_EXEC_BIT=true \
+		-e ENABLE_GITHUB_ACTIONS_GROUP_TITLE=true \
+		-e DEFAULT_BRANCH=main \
+		-e IGNORE_GENERATED_FILES=true \
+		-e IGNORE_GITIGNORED_FILES=true \
+		-e VALIDATE_ALL_CODEBASE=false \
```
Comment (ferrarimarco) [inline Makefile:156]: Can you set this to `true`?

**Review by ferrarimarco — CHANGES_REQUESTED** [review]

**Review by ferrarimarco — APPROVED** [review]
@kftsehk thanks for this PR. I've updated it to address review comments, and should be good to go. Let's wait for the CI jobs to complete.

**Review by ferrarimarco — APPROVED** [review]

**Pull request merged**